brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+# Functional Operations
+
+The SOMA API includes functional capabilities around the [SOMA data model](#datamodel.md). The specifics of how these operations manifest in any given language and computing environment is defined in per-language specifications (**to be created**), but each implementation will minimally support the functional operations defined here. For example, it is likely that a Python implementation will prefer `__getattr__` over an explicit `get()` method, and will add other Pythonic functionality to facilitate use.
+
+As any given storage "engine" upon which SOMA is implemented may have additional features and capabilities required to support advanced use cases, it is expected that SOMA implementations will expose engine-specific features. Where possible, these should be implemented to avoid conflict with future changes to the common SOMA API.
+
+> ⚠️ **Issue** - this spec needs to provide guidance on HOW to do that. Fro example, can we carve out some namespace for extensions in the functional API?
+
+In the following, a pseudo-object-oriented syntax is used, eg, `obj.op()` is calling `op()` upon `obj`. Where a type is used, eg, SOMA.op(), it is a `static` operation (ie, no state other than parameters provided).
+
+> ℹ️ **Note** - this section is just a sketch. Primarily focused on adding pseudo-signatures for any primitive operation that must exist on each type.
+
+## SOMAMapping interface
+
+> ℹ️ **Note** - this is an interface defined only to make the subsequent prose simpler. It is not a primary element in the data model.
+
+The SOMAMapping is a string-keyed map. It may be immutable or mutable depending on the context. In most implementations, it will be presented with a language-appropriate interface, eg, Python `Mapping` or `MutableMapping`.
+
+The following operations manipulate the mapping, providing a getter/setter interface plus the ability to iterate on the collection:
+
+```
+  mapping.get(string key) -> value
+  mapping.has(string key) -> bool
+  mapping.set(string key, simple value) -> void
+  mapping.del(string key) -> void
+  mapping.__iterator__
+  mapping.__length__
+```
+
+> ⚠️ Issue - do we need a "is_readonly" or "is_immutable" attribute, or assume user can figure it out from the context?
+
+> ℹ️ **Note** - it is possible that the data model will grow to include more complex value types. If possible, retain that future option in any API defined.
+
+## SOMACollection
+
+In the following, `soco` refers to an SOMACollection instance or handle.
```

Comment:
as above, `SOMACollection` has two roles -- 'outer' (contains SOMAs) and 'inner' (contains elements within a SOMA, e.g. `obsm`/`varm`)

i think these ops are necessary for both roles

Reply:
agree, and that was my intent by using the same type name in both cases, with the same semantics, operations, etc.

Did you see a place where this was incorrectly documented, or is a request to add more explicit prose?